packages/client/hmi-client/src/workflow/ops/model-config/tera-model-config.vue

```diff
@@ -143,16 +180,134 @@ const emit = defineEmits([
+let editor: VAceEditorInstance['_editor'] | null;
+const buildJupyterContext = () => {
+	const contextId = selectedConfigId.value ?? props.node.state.tempConfigId;
+	if (!model.value) {
+		logger.warn('Cannot build Jupyter context without a model');
+		return null;
+	}
+	return {
+		context: 'mira_config_edit',
+		language: 'python3',
+		context_info: {
+			id: contextId
+		}
+	};
+};
+const codeText = ref(
+	'# This environment contains the variable "model_config" to be read and updated'
+);
+const notebookResponse = ref();
+const sampleAgentQuestions = [
+	'What are the current parameters values?',
+	'update the parameters {gamma: 0.13}'
+];
+
+const appendCode = (data: any, property: string, runUpdatedCode = false) => {
+	codeText.value = codeText.value.concat(' \n', data.content[property] as string);
+	if (runUpdatedCode) runFromCode();
+};
+
+const runFromCode = () => {
+	const code = editor?.getValue();
+	if (!code) return;
+
+	const messageContent = {
+		silent: false,
+		store_history: false,
+		user_expressions: {},
+		allow_stdin: true,
+		stop_on_error: false,
+		code
+	};
+
+	let executedCode = '';
+
+	kernelManager
+		.sendMessage('execute_request', messageContent)
+		.register('execute_input', (data) => {
+			executedCode = data.content.code;
+		})
+		.register('stream', (data) => {
+			notebookResponse.value = data.content.text;
+		})
+		.register('error', (data) => {
+			logger.error(`${data.content.ename}: ${data.content.evalue}`);
+			console.log('error', data.content);
+		})
+		.register('model_preview', (data) => {
+			if (!data.content) return;
+			handleModelPreview(data);
+
+			if (executedCode) {
+				saveCodeToState(executedCode, true);
+			}
+		});
+};
+
+const saveCodeToState = (code: string, hasCodeBeenRun: boolean) => {
+	const state = _.cloneDeep(props.node.state);
+	state.hasCodeBeenRun = hasCodeBeenRun;
+
+	// for now only save the last code executed, may want to save all code executed in the future
+	const codeHistoryLength = props.node.state.modelEditCodeHistory.length;
+	const timestamp = Date.now();
+	if (codeHistoryLength > 0) {
+		state.modelEditCodeHistory[0] = { code, timestamp };
+	} else {
+		state.modelEditCodeHistory.push({ code, timestamp });
+	}
+	emit('update-state', state);
+};
+
+const initializeEditor = (editorInstance: any) => {
+	editor = editorInstance;
+};
+
+const handleModelPreview = (data: any) => {
+	if (!model.value) return;
+	// Only update the keys provided in the model preview (not ID, temporary ect)
+	Object.assign(model.value, _.cloneDeep(data.content['application/json']));
+
+	knobs.value.initials =
+		model.value?.semantics?.ode.initials !== undefined ? model.value?.semantics?.ode.initials : [];
+	knobs.value.parameters =
+		model.value?.semantics?.ode.parameters !== undefined
+			? model.value?.semantics?.ode.parameters
+			: [];
+	knobs.value.timeseries =
+		model.value?.metadata?.timeseries !== undefined ? model.value?.metadata?.timeseries : [];
+};
```

Comment:
Can we show Demeter some love and not have the same chain repeated over and over?

```
const ode = model.value?.semantics?.ode;
if (ode) return;

knobs.value.initials  = ...
knobs.value.parameters = ...

```

Reply:
I still need an individual check for each parameter and initials as theyre both potentially undefined within `ode` but yes I can use `const ode` to shorten the length of nonsense here